File: oxalia_back/app/models/exam.py
import enum
import uuid

from sqlalchemy import Enum, ForeignKey, Integer, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.core.mixins import TimestampMixin, UUIDMixin
from app.database import Base


class ExamStatus(str, enum.Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class Exam(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "exams"

    owner_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    original_filename: Mapped[str] = mapped_column(String(255), nullable=False)
    storage_path: Mapped[str] = mapped_column(String(500), nullable=False)
    content_type: Mapped[str] = mapped_column(String(100), nullable=False)
    size_bytes: Mapped[int] = mapped_column(Integer, nullable=False)
    status: Mapped[ExamStatus] = mapped_column(
        Enum(ExamStatus, name="exam_status", native_enum=False),
        nullable=False,
        default=ExamStatus.PENDING,
    )

    def __repr__(self) -> str:
        return f"<Exam id={self.id} status={self.status}>"

File: oxalia_front/lib/data/models/exam.dart
/// Mirrors the backend `ExamOut` schema and `ExamStatus` enum.
enum ExamStatus {
  pending,
  processing,
  completed,
  failed;

  static ExamStatus fromString(String value) {
    return ExamStatus.values.asNameMap()[value] ?? ExamStatus.pending;
  }
}

class Exam {
  const Exam({
    required this.id,
    required this.originalFilename,
    required this.contentType,
    required this.sizeBytes,
    required this.status,
    required this.createdAt,
  });

  final String id;
  final String originalFilename;
  final String contentType;
  final int sizeBytes;
  final ExamStatus status;
  final DateTime createdAt;

  factory Exam.fromJson(Map<String, dynamic> json) {
    return Exam(
      id: json['id'] as String,
      originalFilename: json['original_filename'] as String,
      contentType: json['content_type'] as String,
      sizeBytes: json['size_bytes'] as int,
      status: ExamStatus.fromString(json['status'] as String),
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }
}
